README.md
# KenBurns

The Ken Burns effect is a type of panning and zooming effect used in video production from still imagery.

First add to your dependencies:

```
dependencies:
  kenburns_nullsafety: 
```

Then import:

```
  import 'package:kenburns_nullsafety/kenburns_nullsafety.dart';
```

Wrap your image with a KenBurns widget
```dart
Container(
      height: 300,
      child: KenBurns(
        child: Image.network("https://lemag.nikonclub.fr/wp-content/uploads/2017/07/08.jpg", fit: BoxFit.cover,),
      ),
),
```

[![screen](https://raw.githubusercontent.com/florent37/Flutter-KenBurns/master/medias/kenburns_slow.gif)](https://github.com/the-Jinxist/kenburns_nullsafety.git)

# Configuration

You can configure KenBurns Widget

```
KenBurns(
    minAnimationDuration : Duration(milliseconds: 3000),
    maxAnimationDuration : Duration(milliseconds: 10000),
    maxScale : 8,
    child: ...
  });
```

# Multiple images

You can display multiple child in KenBurns with a CrossFade animation

```dart
Container(
    height: 300,
    child: KenBurns.multiple(
      childLoop: 3,
      children: [
        Image.network(
          "https://www.photo-paysage.com/?file=pic_download_link/picture&pid=3100",
          fit: BoxFit.cover,
        ),
        Image.network(
          "https://cdn.getyourguide.com/img/location_img-59-1969619245-148.jpg",
          fit: BoxFit.cover,
        ),
        Image.network(
          "https://www.theglobeandmail.com/resizer/vq3O7LI3hvsjTP2N0m9NwU4W3Eg=/1500x0/filters:quality(80)/arc-anglerfish-tgam-prod-tgam.s3.amazonaws.com/public/4ETF3GZR3NA3RDDW23XDRBKKCI",
          fit: BoxFit.cover,
        ),
      ],
    ),
),
```


CHANGELOG.md
## 1.0.0

* Migration to null-safe

## 1.0.1

* Added linter, removed warnings
* Removed unfinished localization 


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:kenburns_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate((Widget widget) {
        return widget is Text &&
            (widget.data?.startsWith('Running on:') == true);
      }),
      findsOneWidget,
    );
  });
}


lib/generated/i18n.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

// ignore_for_file: non_constant_identifier_names
// ignore_for_file: camel_case_types
// ignore_for_file: prefer_single_quotes

// This file is automatically generated. DO NOT EDIT, all your changes would be lost.
class S implements WidgetsLocalizations {
  const S();

  static S? current;

  static const GeneratedLocalizationsDelegate delegate =
      GeneratedLocalizationsDelegate();

  static S? of(BuildContext context) => Localizations.of<S>(context, S);

  @override
  TextDirection get textDirection => TextDirection.ltr;
}

class $en extends S {
  const $en();
}

class GeneratedLocalizationsDelegate extends LocalizationsDelegate<S?> {
  const GeneratedLocalizationsDelegate();

  List<Locale> get supportedLocales {
    return const <Locale>[
      Locale("en", ""),
    ];
  }

  LocaleListResolutionCallback listResolution(
      {Locale? fallback, bool withCountry = true}) {
    return (List<Locale>? locales, Iterable<Locale> supported) {
      if (locales == null || locales.isEmpty) {
        return fallback ?? supported.first;
      } else {
        return _resolve(locales.first, fallback, supported, withCountry);
      }
    };
  }

  LocaleResolutionCallback resolution(
      {Locale? fallback, bool withCountry = true}) {
    return (Locale? locale, Iterable<Locale> supported) {
      return _resolve(locale, fallback, supported, withCountry);
    };
  }

  @override
  Future<S?> load(Locale locale) {
    final String? lang = getLang(locale);
    if (lang != null) {
      switch (lang) {
        case "en":
          S.current = const $en();
          return SynchronousFuture<S?>(S.current);
        default:
        // NO-OP.
      }
    }
    S.current = const S();
    return SynchronousFuture<S?>(S.current);
  }

  @override
  bool isSupported(Locale locale) => _isSupported(locale, true);

  @override
  bool shouldReload(GeneratedLocalizationsDelegate old) => false;

  ///
  /// Internal method to resolve a locale from a list of locales.
  ///
  Locale _resolve(Locale? locale, Locale? fallback, Iterable<Locale> supported,
      bool withCountry) {
    if (locale == null || !_isSupported(locale, withCountry)) {
      return fallback ?? supported.first;
    }

    final Locale languageLocale = Locale(locale.languageCode, "");
    if (supported.contains(locale)) {
      return locale;
    } else if (supported.contains(languageLocale)) {
      return languageLocale;
    } else {
      final Locale fallbackLocale = fallback ?? supported.first;
      return fallbackLocale;
    }
  }

  ///
  /// Returns true if the specified locale is supported, false otherwise.
  ///
  bool _isSupported(Locale locale, bool withCountry) {
    for (Locale supportedLocale in supportedLocales) {
      // Language must always match both locales.
      if (supportedLocale.languageCode != locale.languageCode) {
        continue;
      }

      // If country code matches, return this locale.
      if (supportedLocale.countryCode == locale.countryCode) {
        return true;
      }

      // If no country requirement is requested, check if this locale has no country.
      if (true != withCountry &&
          (supportedLocale.countryCode == null ||
              supportedLocale.countryCode!.isEmpty)) {
        return true;
      }
    }
    return false;
  }
}

String? getLang(Locale l) => l == null
    ? null
    : l.countryCode != null && l.countryCode!.isEmpty
        ? l.languageCode
        : l.toString();


lib/KenburnsGenerator.dart
import 'dart:math';
import 'dart:ui';

import 'package:flutter/foundation.dart';

/// The generated configuration of KenBurns (scale, translation, duration)
class KenBurnsGeneratorConfig {
  double newScale;
  Offset newTranslation;
  Duration newDuration;

  KenBurnsGeneratorConfig({
    required this.newScale,
    required this.newTranslation,
    required this.newDuration,
  });
}

/// The random scale, translation, duration generator
class KenburnsGenerator {
  Random _random = Random();

  KenburnsGenerator();

  /// Generates a positive random integer distributed on the range
  double _randomValue(double min, double max) =>
      min + _random.nextDouble() * (max - min);

  double generateNextScale(
      {double? lastScale, double? maxScale, required bool scaleDown}) {
    final double minScale = 1.0;
    if (scaleDown && minScale < lastScale!) {
      return _randomValue(minScale, lastScale);
    } else {
      return _randomValue(max(minScale, lastScale!), maxScale!);
    }
  }

  Duration generateNextDuration(
      {required double minDurationMillis, required double maxDurationMillis}) {
    return Duration(
        milliseconds:
            _randomValue(minDurationMillis, maxDurationMillis).floor());
  }

  Offset generateNextTranslation(
      {required double width, required double height, required Size nextSize, double? nextScale}) {
    final availableXOffset = ((nextSize.width - width) / 2);
    final availableYOffset = ((nextSize.height - height) / 2);

    final x = _randomValue(-1 * availableXOffset, availableXOffset);
    final y = _randomValue(-1 * availableYOffset, availableYOffset);
    return Offset(x, y);
  }

  KenBurnsGeneratorConfig generateNextConfig({
    required double width,
    required double height,
    double? maxScale,
    double? lastScale,
    required bool scaleDown,
    required double minDurationMillis,
    required double maxDurationMillis,
    Offset? lastTranslation,
  }) {
    Duration nextDuration;
    double nextScale;
    Offset nextTranslation;

    nextDuration = generateNextDuration(
      minDurationMillis: minDurationMillis,
      maxDurationMillis: maxDurationMillis,
    );

    nextScale = generateNextScale(
      lastScale: lastScale,
      maxScale: maxScale,
      scaleDown: scaleDown,
    );

    Size nextSize = Size(width * nextScale, height * nextScale);

    nextTranslation = generateNextTranslation(
      width: width,
      height: height,
      nextScale: nextScale,
      nextSize: nextSize,
    );

    return KenBurnsGeneratorConfig(
      newDuration: nextDuration,
      newTranslation: nextTranslation,
      newScale: nextScale,
    );
  }
}


lib/kenburns_generator.dart
import 'dart:math';
import 'dart:ui';

/// The generated configuration of KenBurns (scale, translation, duration)
class KenBurnsGeneratorConfig {
  double newScale;
  Offset newTranslation;
  Duration newDuration;

  KenBurnsGeneratorConfig({
    required this.newScale,
    required this.newTranslation,
    required this.newDuration,
  });
}

/// The random scale, translation, duration generator
class KenburnsGenerator {
  final Random _random = Random();

  KenburnsGenerator();

  /// Generates a positive random integer distributed on the range
  double _randomValue(double min, double max) =>
      min + _random.nextDouble() * (max - min);

  double generateNextScale(
      {double? lastScale, double? maxScale, required bool scaleDown}) {
    const double minScale = 1.0;
    if (scaleDown && minScale < lastScale!) {
      return _randomValue(minScale, lastScale);
    } else {
      return _randomValue(max(minScale, lastScale!), maxScale!);
    }
  }

  Duration generateNextDuration(
      {required double minDurationMillis, required double maxDurationMillis}) {
    return Duration(
        milliseconds:
            _randomValue(minDurationMillis, maxDurationMillis).floor());
  }

  Offset generateNextTranslation(
      {required double width,
      required double height,
      required Size nextSize,
      double? nextScale}) {
    final availableXOffset = ((nextSize.width - width) / 2);
    final availableYOffset = ((nextSize.height - height) / 2);

    final x = _randomValue(-1 * availableXOffset, availableXOffset);
    final y = _randomValue(-1 * availableYOffset, availableYOffset);
    return Offset(x, y);
  }

  KenBurnsGeneratorConfig generateNextConfig({
    required double width,
    required double height,
    double? maxScale,
    double? lastScale,
    required bool scaleDown,
    required double minDurationMillis,
    required double maxDurationMillis,
    Offset? lastTranslation,
  }) {
    Duration nextDuration;
    double nextScale;
    Offset nextTranslation;

    nextDuration = generateNextDuration(
      minDurationMillis: minDurationMillis,
      maxDurationMillis: maxDurationMillis,
    );

    nextScale = generateNextScale(
      lastScale: lastScale,
      maxScale: maxScale,
      scaleDown: scaleDown,
    );

    Size nextSize = Size(width * nextScale, height * nextScale);

    nextTranslation = generateNextTranslation(
      width: width,
      height: height,
      nextScale: nextScale,
      nextSize: nextSize,
    );

    return KenBurnsGeneratorConfig(
      newDuration: nextDuration,
      newTranslation: nextTranslation,
      newScale: nextScale,
    );
  }
}


lib/kenburns_nullsafety.dart
import 'package:flutter/material.dart';

import 'kenburns_generator.dart';

/// KenBurns widget, please provide a `child` Widget,
/// Will animate the child, using random scale, translation & duration
class KenBurns extends StatefulWidget {
  final Widget? child;

  /// minimum translation & scale duration, not null
  final Duration minAnimationDuration;

  /// maximum translation & scale duration, not null
  final Duration maxAnimationDuration;

  /// Maximum allowed child scale, > 1
  final double maxScale;

  //region multiple images
  /// If specified (using the constructor multiple)
  /// Will animate [childLoop] each children then will fade to the next child
  /// Not Null & Size must be > 1
  /// if size == 1 -> Will use the KenBurns as a single child
  final List<Widget>? children;

  /// If specified (using the constructor multiple)
  /// Will specify the fade in duration between 2 child
  final Duration? childrenFadeDuration;

  /// If specified (using the constructor multiple)
  /// Will determine how many times each child will stay in the KenBurns
  /// Until the next child will be displayed
  final int? childLoop;

  //endregion

  /// Constructor for a single child KenBurns
  KenBurns({
    Key? key,
    required Widget this.child,
    this.minAnimationDuration = const Duration(milliseconds: 3000),
    this.maxAnimationDuration = const Duration(milliseconds: 10000),
    this.maxScale = 8,
  })  : childrenFadeDuration = null,
        children = null,
        childLoop = null,
        assert(minAnimationDuration.inMilliseconds > 0),
        assert(maxAnimationDuration.inMilliseconds > 0),
        assert(minAnimationDuration < maxAnimationDuration),
        assert(maxScale > 1),
        super(key: key);

  /// Constructor for multiple child KenBurns
  const KenBurns.multiple(
      {Key? key,
      this.minAnimationDuration = const Duration(milliseconds: 1000),
      this.maxAnimationDuration = const Duration(milliseconds: 10000),
      this.maxScale = 10,
      this.childLoop = 3,
      this.children,
      this.childrenFadeDuration = const Duration(milliseconds: 800)})
      : child = null,
        super(key: key);

  @override
  State<KenBurns> createState() => _KenBurnsState();
}

class _KenBurnsState extends State<KenBurns> with TickerProviderStateMixin {
  bool _running = false;

  /// The generated scale controller
  /// Will be destroyed / created at each loop (because duration is different)
  AnimationController? _scaleController;

  /// The generated scale controller's animation
  /// Will be destroyed / created at each loop (because duration is different)
  late Animation<double> _scaleAnim;

  /// The generated translation controller
  /// Will be destroyed / created at each loop (because duration is different)
  AnimationController? _translationController;

  /// The generated translation controller's X animation
  /// Will be destroyed / created at each loop (because duration is different)
  late Animation<double> _translationXAnim;

  /// The generated translation controller's Y animation
  /// Will be destroyed / created at each loop (because duration is different)
  late Animation<double> _translationYAnim;

  /// The animated current scale
  double _currentScale = 1;

  /// The animated current translation X
  double _currentTranslationX = 0;

  /// The animated current translation Y
  double _currentTranslationY = 0;

  /// If true : next animation will scale down,
  /// false : next animation will scale up
  bool _scaleDown = true;

  /// For developpers : set to true to enable logs
  final bool _displayLogs = false;

  /// The random [scale/duration/translation] generator
  final KenburnsGenerator _kenburnsGenerator = KenburnsGenerator();

  //region multiple childs
  /// if true : the widget setup is multipleImages
  bool get _displayMultipleImage =>
      widget.children != null && widget.children!.length > 1;
  int _nextChildIndex = -1;
  int _currentChildIndex = 0;
  int _currentChildLoop = 0;

  double _opacityCurrentChild = 1;
  double _opacityNextChild = 0;

  /// The generated fade controller
  AnimationController? _fadeController;

  /// The generated opacity fade in controller's animation
  late Animation<double> _fadeInAnim;

  /// The generated opacity fade out controller's animation
  late Animation<double> _fadeOutAnim;

  //endregion

  /// Generate the fade (in & out) animations
  Future<void> _createFadeAnimations() async {
    _fadeController?.dispose();
    _fadeController = AnimationController(
      duration: widget.childrenFadeDuration,
      vsync: this,
    );
    _fadeInAnim = Tween(begin: 0.0, end: 1.0).animate(
      CurvedAnimation(parent: _fadeController!, curve: Curves.linear),
    )..addListener(() {
        setState(() {
          _opacityNextChild = _fadeInAnim.value;
        });
      });
    _fadeOutAnim = Tween(begin: 1.0, end: 0.0).animate(
      CurvedAnimation(parent: _fadeController!, curve: Curves.linear),
    )..addListener(() {
        setState(() {
          _opacityCurrentChild = _fadeOutAnim.value;
        });
      });
  }

  /// Generate the next animation [scale, duration, translation]
  /// Using the [KenBurnsGenerator] generateNextConfig
  Future<void> _createNextAnimations(
      {required double height, required double width}) async {
    final KenBurnsGeneratorConfig nextConfig =
        _kenburnsGenerator.generateNextConfig(
            width: width,
            height: height,
            maxScale: widget.maxScale,
            lastScale: _currentScale,
            scaleDown: _scaleDown,
            minDurationMillis:
                widget.minAnimationDuration.inMilliseconds.toDouble(),
            maxDurationMillis:
                widget.maxAnimationDuration.inMilliseconds.toDouble(),
            lastTranslation:
                Offset(_currentTranslationX, _currentTranslationY));

    /// Recreate the scale animations
    _scaleController?.dispose();
    _scaleController = AnimationController(
      duration: nextConfig.newDuration,
      vsync: this,
    );

    _scaleAnim = Tween(begin: _currentScale, end: nextConfig.newScale).animate(
      CurvedAnimation(parent: _scaleController!, curve: Curves.linear),
    )..addListener(() {
        setState(() {
          _currentScale = _scaleAnim.value;
        });
      });

    /// Recreate the translations animations
    _translationController?.dispose();
    _translationController = AnimationController(
      duration: nextConfig.newDuration,
      vsync: this,
    );

    _translationXAnim =
        Tween(begin: _currentTranslationX, end: nextConfig.newTranslation.dx)
            .animate(
      CurvedAnimation(parent: _translationController!, curve: Curves.linear),
    )..addListener(() {
            setState(() {
              _currentTranslationX = _translationXAnim.value;
            });
          });
    _translationYAnim =
        Tween(begin: _currentTranslationY, end: nextConfig.newTranslation.dy)
            .animate(
      CurvedAnimation(parent: _translationController!, curve: Curves.linear),
    )..addListener(() {
            setState(() {
              _currentTranslationY = _translationYAnim.value;
            });
          });

    log("kenburns started");
    log("kenburns d(${nextConfig.newDuration}) translation(${nextConfig.newTranslation.dx}, ${nextConfig.newTranslation.dy}) scale(${nextConfig.newScale})");

    /// Next scale animation will be inverted
    _scaleDown = !_scaleDown;

    /// fire scale & translation animations
    await Future.wait(
        [_scaleController!.forward(), _translationController!.forward()]);

    log("kenburns finished");
  }

  /// Display on debug logs (enable with [_displayLogs])
  void log(String text) {
    if (_displayLogs) {
      log(text);
    }
  }

  /// Fire the fade (in/out) animation
  Future<void> _fade() async {
    await _fadeController!.forward();

    if (!_running) return;

    setState(() {
      _currentChildIndex = _nextChildIndex;

      _nextChildIndex = _currentChildIndex + 1;
      _nextChildIndex = _nextChildIndex % widget.children!.length;
    });

    _fadeController!.reset();
  }

  Future<void> fire({double? height, double? width}) async {
    _running = true;
    if (_displayMultipleImage) {
      _nextChildIndex = 1;

      /// Create one time the fade animation
      await _createFadeAnimations();

      /// Cancel if _running go to false
      while (_running) {
        await _createNextAnimations(width: width!, height: height!);
        if (!_running) return;

        if (_currentChildLoop % widget.childLoop! == 0) {
          _fade(); //parallel
        }
        _currentChildLoop++;
      }
    } else {
      /// Cancel if _running go to false
      while (_running) {
        await _createNextAnimations(width: width!, height: height!);
      }
    }
  }

  @override
  void initState() {
    /// Reset _runnint state
    _running = false;
    super.initState();
  }

  @override
  void didUpdateWidget(KenBurns oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (oldWidget.children?.length != oldWidget.children?.length) {
      _running = false;
      _scaleController?.dispose();
      _fadeController?.dispose();
      _translationController?.dispose();
      _currentChildIndex = 0;
    }
  }

  @override
  Widget build(BuildContext context) {
    /// Layout builder to provide constrains height/width
    return LayoutBuilder(builder: (context, constraints) {
      /// create the animation only if we have a size (not possible in initState())
      if (!_running) {
        fire(height: constraints.maxHeight, width: constraints.maxWidth);
      }
      return ClipRect(
        ///Clip because we scale up children, if not clipped : child can take all the screen
        /// Apply the current animated translation
        child: Transform.translate(
          offset: Offset(_currentTranslationX, _currentTranslationY),

          /// Apply the current animated scale
          child: Transform.scale(
            scale: _currentScale,
            child: _buildChild(),
          ),
        ),
      );
    });
  }

  Widget _buildChild() {
    if (_displayMultipleImage) {
      /// If the [currentChildIndex] changed (different than [lastChildIndex])
      /// -> we animate to display the next child
      /// We use the stack to keep the same structure as multiple/single child
      return Stack(fit: StackFit.expand, children: <Widget>[
        Opacity(
            opacity: _opacityCurrentChild,
            child: widget.children![_currentChildIndex]),
        Opacity(
            opacity: _opacityNextChild,
            child: widget.children![_nextChildIndex]),
      ]);
    } else {
      /// If we have only 1 child
      /// We use the stack to keep the same structure as multiple/single child
      return Stack(
        fit: StackFit.expand,
        children: [
          widget.child!,
        ],
      );
    }
  }

  @override
  void dispose() {
    /// will stop the [fire()] loop
    _running = false;
    _scaleController?.dispose();
    _translationController?.dispose();
    _fadeController?.dispose();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:kenburns_nullsafety/kenburns_nullsafety.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Padding(
          padding: const EdgeInsets.all(30.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Container(
                height: 300,
                child: KenBurns.multiple(
                  maxAnimationDuration: Duration(seconds: 10),
                  minAnimationDuration: Duration(seconds: 3),
                  children: <Widget>[
                    Image.network(
                      "https://cdn.hasselblad.com/hasselblad-com/6cb604081ef3086569319ddb5adcae66298a28c5_x1d-ii-sample-01-web.jpg?auto=format&q=97",
                      fit: BoxFit.cover,
                    ),
                    Image.network(
                      "https://cdn.getyourguide.com/img/location_img-59-1969619245-148.jpg",
                      fit: BoxFit.cover,
                    ),
                    Image.network(
                      "https://www.theglobeandmail.com/resizer/vq3O7LI3hvsjTP2N0m9NwU4W3Eg=/1500x0/filters:quality(80)/arc-anglerfish-tgam-prod-tgam.s3.amazonaws.com/public/4ETF3GZR3NA3RDDW23XDRBKKCI",
                      fit: BoxFit.cover,
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
